app/javascript/components/widget-display.jsx

```diff
@@ -20,18 +21,111 @@ const widgetElements = {
   Traffic,
 };
 
+const prefetchDelay = 1500;
+let lastPrefetchTime = 0;
+
+const isPaused = state => state.start === null;
+
+const reducer = (state, action) => {
+  const { start, durationRemaining } = state;
+  const now = Date.now();
+
+  switch (action.type) {
+    case 'togglePause':
+      return {
+        durationRemaining: isPaused(state)
+          ? durationRemaining
+          : durationRemaining - (now - start),
+        start: isPaused(state) ? now : null,
+      };
+    case 'startTimer':
+      return {
+        durationRemaining: action.durationSeconds,
+        start: isPaused(state) ? null : now,
+      };
+    default:
+      return state;
+  }
+};
+
 export const WidgetDisplay = ({
-  keyToWidgetAction,
-  clickToWidgetAction,
+  requestPrefetch,
+  requestWidget,
   cityName,
   widgets,
-  selectedWidgetId,
-  widgetName,
-  showWeather,
-  totalTime,
-  isPaused,
 }) => {
-  const WidgetElement = widgetElements[widgetName].Panel;
+  const { byIdOrFirst: current, enabled: enabledWidgets } = widgets;
+
+  const widgetTransitionTime = current.durationSeconds * 1000;
+
+  const WidgetElement = widgetElements[current.name].Panel;
+
+  const index = enabledWidgets.findIndex(w => w.id === current.id);
+
+  const prevWidgetId =
+    enabledWidgets[mathMod(index - 1, enabledWidgets.length)].id;
+  const nextWidgetId =
+    enabledWidgets[mathMod(index + 1, enabledWidgets.length)].id;
+
+  const [state, dispatch] = useReducer(reducer, {
+    start: widgetTransitionTime,
+    durationRemaining: null,
+  });
+
+  useEffect(() => {
+    dispatch({
+      type: 'startTimer',
+      durationSeconds: widgetTransitionTime,
+    });
+  }, [current.id]);
+
+  useEffect(() => {
+    const timerId = setInterval(() => {
+      const { durationRemaining, start } = state;
+      if (isPaused(state)) {
+        return;
+      }
+
+      const prefetchTime = durationRemaining - prefetchDelay;
+      const now = Date.now();
```

Comment:
Something is going wrong and I suspect it stems from here. I noticed the rails server logs were getting more query requests for `getWidgets` after this change. Adding some logging:
```
diff --git a/app/javascript/components/widget-controller.jsx b/app/javascript/components/widget-controller.jsx
index 7715697..9dd5df6 100644
--- a/app/javascript/components/widget-controller.jsx
+++ b/app/javascript/components/widget-controller.jsx
@@ -35,6 +35,7 @@ export const WidgetController = ({ client, cityName }) => {
   const fetchPolicy = newWidgetId === 0 ? 'network-only' : 'cache-first';
 
   const queryCallback = nextWidgetId => {
+    console.log('caling getWidgets');
     client.query({
       query: getWidgets,
       variables: {
```
Shows a console log about every second fetching the widgets. Looking at the changes between the previous logic and here it appears we have switched to ms durations in our state where before we had Date.now() values which is an epoch style numeric time.

Looking at the names `start` and `durationRemaining` is a little misleading. Start actually starts with the ms remaining for the transition in the logic.

We could keep the original state and corresponding logic for this refactor. Or we could delve a little deeper extracting functions from the original logic with some tests coverage to refactor prior to class component to hooks.

Reply:
@cbarber good catch - I cleaned it up a bit and fixed the bug. It looks like I forgot to alter the prefetch logic in the interval to work with the changes made to the timer state. It works now, and lastPrefetchTime isn't necessary anymore which is a bonus.

Regarding the naming of `start`, I can see why it was confusing: I accidentally gave it the `widgetTransitionTime` in the initial state when it should have been `Date.now()`. I also renamed `start` to be `startTimestamp` for less confusion.

Let me know what you think of the changes.